### PR TITLE
Add option to select different datasets and fid fields as input to OQT

### LINKED ISSUE #4: Enable different datasets as input to OQT
**Currently** it is only possible to use the custom regions as input dataset.  For those regions the feature ids is the unique identifier and also the parameter name. E.g. `oqt create-indicator --dataset regions --feature-id 1`

**Expected** is that it should be possible to choose a different dataset like GADM or ISEA (hexagonal cells).
For GADM there are multiple unique IDs such as country ISO3 codes (https://gadm.org/metadata.html). In addition it would be quite nice to be able to use english country names as identifier as well.

**Solution:** I propose to use the parameters 'dataset' and 'id' to identify single features of a specific dataset. Every dataset will have a default unique id field associated with it (fid in case of the oqt-regions, GID_0 (ISO3) in case of GADM). This mapping could be defined as dictionary in `definitions.py`. E.g. `oqt create-indicator --dataset gadm --id deu`

In addition we could provide a way to change the id field which is used by OQT (E.g. to NAME_0 in case of GADM). This could be archived by either adding a new option to CLI (`oqt create-indicator --dataset gadm --id-field NAME_0 --id Germany`) or changing the syntax of the given id (`create-indicator --dataset gadm -id "NAME_0:Germany"`).

### PR DESCRIPTION
**Pull request by matthiasschaub**
### Description

Add option to select different datasets and fid fields as input to OQT.

Available datasets and possible fid fields are defined in `definitions.py`.

Each dataset defaults to a primary fid field. The usage of other fields as identifier is possible.
The attribute/parameter/argument `fid_field`/`fidField` is added to the report class, oqt functions and API/CLI.

Rewrite save and load indicator from database logic to use one result table for all indicator results.

### Corresponding issue
Closes #4 

### Checklist
- [x] I have updated my branch to `main` (e.g. through `git rebase main`)
- [x] My code follows the [style guide](https://github.com/GIScience/ohsome-quality-analyst/blob/main/CONTRIBUTING.md#style-guide) and was checked with [pre-commit](https://github.com/GIScience/ohsome-quality-analyst/blob/main/CONTRIBUTING.md#pre-commit) before committing
- [x] I have commented my code
- [x] I have added sufficient unit and integration [tests](https://github.com/GIScience/ohsome-quality-analyst/blob/main/docs/development_setup.md#tests)
- [x] I have updated the [CHANGELOG.md](https://github.com/GIScience/ohsome-quality-analyst/blob/main/CHANGELOG.md)
- [x] Check changes to VCR cassetts
- [x] Add sensible commits (squash, rebase, …)

### CODE REVIEW

File: workers/ohsome_quality_analyst/cli_opts.py [hunk 1]
```diff
@@ -5,11 +5,28 @@
...
+    https://click.palletsprojects.com/en/8.0.x/parameters/#implementing-custom-types
+    """
+
+    def convert(self, value, param, ctx):
+        try:
+            return int(value)
+        except ValueError:
+            if isinstance(value, str):
+                return value
+        self.fail(f"{0} is not a valid integer or string".format(value), param, ctx)
```
Comment (joker234) [inline workers/ohsome_quality_analyst/cli_opts.py:27]: Do we need a unittest for this class/line?

File: workers/ohsome_quality_analyst/geodatabase/create_results_table.sql [hunk 2]
```diff
@@ -0,0 +1,13 @@
+CREATE TABLE IF NOT EXISTS results (
+    indicator text,
```
Comment (joker234) [inline workers/ohsome_quality_analyst/geodatabase/create_results_table.sql:2]: Be careful with the word `INDICATOR` as this seems to be a reserved word in SQL (although I'm not sure about PostgreSQL). If you need more inside about why and how PostgreSQL deals with SQL-reserved-words: https://www.postgresql.org/docs/current/sql-keywords-appendix.html

P.S.: the same applies to `VALUE`

File: workers/ohsome_quality_analyst/geodatabase/client.py [hunk 3]
```diff
@@ -188,19 +142,20 @@ async def get_area_of_bpolys(bpolys: Dict):
...
 
 
 async def get_bpolys_from_db(
-    dataset: str, feature_id: int
+    dataset: str, feature_id: Union[int, str], fid_field: str
 ) -> geojson.FeatureCollection:
-    """Get geometry and properties from geo database as a geojson feature collection."""
-    logging.info("Get bpolys geometry")
-    # TODO: adjust this for other input tables
+    """Get bounding polygon from the Geodatabase as a GeoJSON Feature Collection."""
```
Comment (joker234) [inline workers/ohsome_quality_analyst/geodatabase/client.py:154]: Either `FeatureCollection` is meant to be as proper name and it's written together, or it's written in small letters :)
```suggestion
    """Get bounding polygon from the geodatabase as a GeoJSON FeatureCollection."""
```

File: workers/ohsome_quality_analyst/geodatabase/client.py [hunk 4]
```diff
@@ -214,33 +169,37 @@ async def get_bpolys_from_db(
...
-                        -- list of fields
-                        'fid', ogc_fid
                     )
                 )
             )
         )
-        FROM {0}
-        WHERE ogc_fid = $1
+        FROM {dataset}
+        WHERE {fid_field} = $1
```
Comment (joker234) [inline workers/ohsome_quality_analyst/geodatabase/client.py:180]: Fortunately, `asyncpg` prevents the most obvious threat. Nevertheless, I would be very careful with taking input directly from the user input to a database query without sanity-checking them. Since this doesn't change in this PR, I would advise to open a high-prio issue for this.

File: workers/ohsome_quality_analyst/geodatabase/client.py [hunk 5]
```diff
@@ -214,33 +169,37 @@ async def get_bpolys_from_db(
...
+    ).format(fid_field=fid_field, dataset=dataset)
+
     async with get_connection() as conn:
-        result = await conn.fetchrow(query, feature_id)
+        try:
+            result = await conn.fetchrow(query, feature_id)
+        except (UndefinedColumnError, DataError):
+            # TODO: Do we need a custom error here?
+            # DataError occurs if feature id is a wrong type. E.g. str not int
+            raise
```
Comment (joker234) [inline workers/ohsome_quality_analyst/geodatabase/client.py:190]: I think a block with `except

File: workers/ohsome_quality_analyst/geodatabase/client.py [hunk 6]
```diff
@@ -214,33 +169,37 @@ async def get_bpolys_from_db(
...
-        query = f.read()
+    working_dir = os.path.dirname(os.path.abspath(__file__))
+    file_path = os.path.join(working_dir, "regions_as_geojson.sql")
+    with open(file_path, "r") as file:
+        query = file.read()
     async with get_connection() as conn:
         record = await conn.fetchrow(query)
     feature_collection = geojson.loads(record[0])
-    # To be complaint with rfc7946 "id" should be a member of Feature
+    # To be complaint with rfc7946 "id" should be a member of Feature not of Properties
```
Comment (joker234) [inline workers/ohsome_quality_analyst/geodatabase/client.py:202]:
```suggestion
    # To be complaint with rfc7946 "id" should be a member of the feature and not of the properties.
```

File: workers/ohsome_quality_analyst/oqt.py [hunk 7]
```diff
@@ -135,15 +147,18 @@ async def create_report(
...
     report_class = name_to_class(class_type="report", name=report_name)
-    report = report_class(bpolys=bpolys, dataset=dataset, feature_id=feature_id)
+    report = report_class(
+        bpolys=bpolys, dataset=dataset, feature_id=feature_id, fid_field=fid_field
+    )
     report.set_indicator_layer()
     for indicator_name, layer_name in report.indicator_layer:
         indicator = await create_indicator(
             indicator_name,
             layer_name,
```
Comment (joker234) [inline workers/ohsome_quality_analyst/oqt.py:157]: Why didn't you change these two to named parameters?

File: workers/tests/integrationtests/test_api_indicator.py [hunk 8]
```diff
@@ -21,9 +21,9 @@ class TestApiIndicator(unittest.TestCase):
     def setUp(self):
         self.client = TestClient(app)
         self.dataset = "regions"
-        self.feature_id = 31
+        self.feature_id = 1
         self.bpolys = asyncio.run(
-            db_client.get_bpolys_from_db(self.dataset, self.feature_id)
+            db_client.get_bpolys_from_db(self.dataset, self.feature_id, "ogc_fid")
```
Comment (joker234) [inline workers/tests/integrationtests/test_api_indicator.py:26]: Why is `"ogc_fid"` hardcoded here and not in a class variable like `self.dataset`? The same question applies to all the other tests.

File: workers/tests/integrationtests/test_api_indicator.py [hunk 9]
```diff
@@ -21,9 +21,9 @@ class TestApiIndicator(unittest.TestCase):
     def setUp(self):
         self.client = TestClient(app)
         self.dataset = "regions"
-        self.feature_id = 31
+        self.feature_id = 1
```
Comment (joker234) [inline workers/tests/integrationtests/test_api_indicator.py:24]: Why did you change the feature *only* for this test? What is different from the others using `31` right now?

File: CHANGELOG.md [hunk 10]
```diff
@@ -2,10 +2,13 @@
 
 ## Current Main
 
+- Add option to select different datasets and fid fields as input to OQT #4 #37
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply CHANGELOG.md:5]: I will change the order to be consistent with past order. Also I will only link to the Issue.

File: workers/ohsome_quality_analyst/base/report.py [hunk 11]
```diff
@@ -40,9 +40,11 @@ def __init__(
         bpolys: FeatureCollection = None,
         dataset: str = None,
         feature_id: int = None,
+        fid_field: str = None,
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/base/report.py:43]: Not realy. The only documentation is provided through the CLI (`cli_opts.py`)

```python
fid_field_opt = [
    click.option(
        "--fid-field",
        type=str,
        help=(
            "Provide the feature id field of the dataset. "
            + "Use command list-fid-fields to view available "
            + "fid fields for each dataset"
        ),
        default=None,
    )
]
```

File: workers/ohsome_quality_analyst/cli.py [hunk 12]
```diff
@@ -210,10 +228,11 @@ def create_report(
         bpolys = None
         report = asyncio.run(
             oqt.create_report(
-                report_name=report_name,
+                report_name,
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/cli.py:231]: My thinking was to make it clear that this is a madatory parameter. All other parameter have a default value definied. Those this makes sense?
But it could be a named parameter also.

```
async def create_report(
    report_name: str,
    force: bool = False,
    bpolys: Optional[FeatureCollection] = None,
    dataset: Optional[str] = None,
    feature_id: Optional[Union[int, str]] = None,
    fid_field: Optional[str] = None,
```

File: workers/ohsome_quality_analyst/cli.py [hunk 13]
```diff
@@ -222,31 +241,20 @@ def create_report(
...
-    ),
-    help=("Choose a dataset containing geometries."),
-)
-def create_all_indicators(dataset_name: str, force: bool):
-    """Create all indicators for a specified dataset."""
+def create_all_indicators(force: bool):
+    """Create all indicators for OQT regions."""
     click.echo(
-        "This command will calculate all indicators for the specified dataset "
+        "This command will calculate all indicators for OQT regions "
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/cli.py:249]: This refers to the custom definied regions of OQT. Those are currently called just regions.

[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/cli.py]: And yes it refers to ALL regions.

File: workers/ohsome_quality_analyst/cli_opts.py (hunk 1 above)
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/cli_opts.py:27]: I added a test case to cover this Class

File: workers/ohsome_quality_analyst/geodatabase/create_results_table.sql (hunk 2 above)
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/geodatabase/create_results_table.sql:2]: Thats a good point.
Two solutions come to mind:
- Quote those identifiers: `"indicator"`
- Rename them to: `indicator_` or `_indicator`
- Rename them to: `indicator_name` and `val`

Do you have a preference here?

File: workers/ohsome_quality_analyst/oqt.py (hunk 7 above)
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/oqt.py:157]: Because they are required by the function `create_indicator` regardless of the other parameters, which have default values.

File: workers/tests/integrationtests/test_api_indicator.py (hunk 8 above)
Comment (matthiasschaub) [reply workers/tests/integrationtests/test_api_indicator.py:26]: I changed it to be consistent with the other code

File: workers/tests/integrationtests/test_api_indicator.py (hunk 9 above)
Comment (matthiasschaub) [reply workers/tests/integrationtests/test_api_indicator.py]: Reverted the change

File: workers/ohsome_quality_analyst/base/report.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/base/report.py:43]: In Commit e89d9e80187fd55b9ddd5ce5c38d78f4251cd13e I made to sanity check more explicit by using functions and also avoid code duplication. Also I added missing sanity checks in api.py.

File: workers/ohsome_quality_analyst/geodatabase/client.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/geodatabase/client.py:192]: Your are right. I removed the tr except clause. Those error should not be encountered anymore since dataset and fid_field are checked beforehand if they are valid

File: workers/ohsome_quality_analyst/geodatabase/client.py (hunk 4 above)
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/geodatabase/client.py:181]: In Commit e89d9e8 I made to sanity check more explicit by using functions and also avoid code duplication. Also I added missing sanity checks in api.py.

File: workers/ohsome_quality_analyst/api.py [hunk 14]
```diff
@@ -173,9 +190,18 @@ async def _fetch_report(
 
 
 @app.get("/geometries/{dataset}")
-async def get_bpolys_from_db(dataset: str, featureId: int):
-    bpolys = await db_client.get_bpolys_from_db(dataset=dataset, feature_id=featureId)
-    return bpolys
+async def get_bpolys_from_db(
```
Comment (matthiasschaub) [inline workers/ohsome_quality_analyst/api.py:193]: Tests are not sufficient for this function. But I will remove the function in a separate PR since it is not in use.

File: workers/ohsome_quality_analyst/geodatabase/client.py [hunk 15]
```diff
@@ -214,33 +170,52 @@ async def get_bpolys_from_db(
...
         feature["id"] = feature["properties"].pop("id")
     return feature_collection
+
+
+def sanity_check_dataset(dataset: str) -> bool:
+    """Compare against pre-definied values to prevent SQL injection"""
+    if dataset not in DATASETS.keys():
+        return False
+    else:
+        return True
```
Comment (joker234) [inline workers/ohsome_quality_analyst/geodatabase/client.py:210]:
```suggestion
    return dataset in DATASETS.keys()
```

File: workers/ohsome_quality_analyst/geodatabase/client.py [hunk 16]
```diff
@@ -214,33 +170,52 @@ async def get_bpolys_from_db(
...
+
+def sanity_check_fid_field(dataset: str, fid_field: str) -> bool:
+    """Compare against pre-definied values to prevent SQL injection"""
+    if (
+        fid_field not in DATASETS[dataset]["other"]
+        and fid_field != DATASETS[dataset]["default"]
+    ):
+        return False
+    else:
+        return True
```
Comment (joker234) [inline workers/ohsome_quality_analyst/geodatabase/client.py:221]:
```suggestion
    return fid_field in DATASETS[dataset]["other"] or fid_field == DATASETS[dataset]["default"]
```

File: workers/ohsome_quality_analyst/geodatabase/client.py (hunk 4 above)
Comment (joker234) [reply workers/ohsome_quality_analyst/geodatabase/client.py:181]: Hmm, with the last reviewed version, I got the asyncpg error that I can't execute multiple operations in one query. So somehow I got through to asyncpg. Now I get the `ValueError` :)

File: workers/ohsome_quality_analyst/geodatabase/create_results_table.sql (hunk 2 above)
Comment (joker234) [reply workers/ohsome_quality_analyst/geodatabase/create_results_table.sql:2]: as the value is representation the indicator name, I would use `indicator_name`.

**Review by joker234 — COMMENTED** [review]
I think you should consider redoing the VCRs with a fresh cassette again. I got some changed/additional requests when running `pytest` with a fresh database.

File: workers/ohsome_quality_analyst/geodatabase/create_results_table.sql (hunk 2 above)
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/geodatabase/create_results_table.sql:2]: I renamed attributes to avoid usage of SQL keywords

**Review by joker234 — APPROVED** [review]
